### PR TITLE
RFC to allow self parameter be any type

### PR DESCRIPTION
**Pull request by KalitaAlexey**
[Rendered](https://github.com/KalitaAlexey/rfcs/blob/1577f265b3a69c6e38d71e23266500b44277229b/text/0000-any_type_self.md)

### CODE REVIEW

File: text/0000-any_type_self.md [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+            sound.play()
+        });  
+    }
+}
+```
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+No drawbacks.
```
Comment (steveklabnik) [inline text/0000-any_type_self.md]: There is never a proposal with no drawbacks.

**Comment by steveklabnik** [comment]
I am not completely opposed to this idea, but this RFC is _extremely_ light on details.

File: text/0000-any_type_self.md [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+
+# Summary
+[summary]: #summary
+
+Allow self parameter be any type.
+
+# Motivation
+[motivation]: #motivation
+
+It allows to avoid wrapper structs.
```
Comment (jonas-schievink) [inline text/0000-any_type_self.md]: An example showing where a wrapper struct is currently required would be nice

**Comment by jonas-schievink** [comment]
What happens when I do this:

```rust
struct MyStruct;

impl MyStruct {
    pub fn len(self: &Vec<MyStruct>) -> usize { ... }
}

fn main() {
    let v: Vec<MyStruct> = Vec::new();
    v.len();   // which method is called?
}
```

Will this shadow `Vec::len()`? Will it not compile?

**Comment by KalitaAlexey** [comment]
@steveklabnik I know but I can't figure out any more details.

**Comment by eddyb** [comment]
The guarantee needs to be stronger, in order to be able to implement this in a meaningful way.
Last time I looked into this, `Deref<Target=Self>` seemed to be enough for inherent methods.
That means that the `Self` type is found when doing recursive dereferences of `

Trait methods, however, if they can be used in trait objects, require that `typeof self` is some `X<Self>` where `X<Self>: CoerceUnsized<X<Trait>>` where `X<Self>` and `X<Trait>` are structurally identical down to a matching pointer/reference to `Self`, and `Trait`, respectively.
Otherwise, associated types could be used to convert a pointer into a reference, in safe code.

cc @rust-lang/lang

**Comment by KalitaAlexey** [comment]
@jonas-schievink It should be ambiguous code error.

**Comment by eddyb** [comment]
@KalitaAlexey @jonas-schievink No, that method can't be in an `impl MyStruct`, at most `impl [MyStruct]`. It *bears no direct relation* to `MyStruct`, only to `[MyStruct]`.

**Comment by KalitaAlexey** [comment]
@eddyb As you wish it is an option also.

File: text/0000-any_type_self.md (hunk 2 above)
Comment (KalitaAlexey) [inline text/0000-any_type_self.md]: @Jonas-schievink See detailed design section.

File: text/0000-any_type_self.md [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+
+# Detailed design
+[design]: #detailed-design
+
+Currently self parameter type is allowed to be only `Self`, `&Self`, `&mut Self`, `Box<Self>`.
+
+Solution: Allow `self` be any type that depends on `Self`.
```
Comment (KalitaAlexey) [inline text/0000-any_type_self.md]: @jonas-schievink Type depends on `Self` = There is `Self` in Type. For example:
`Vec<Self>`, `Box<Self>, `MyStruct<Self>`.

**Comment by nrc** [comment]
I believe this has previously been discussed in the context of UFCS, it would be good to dig up and reference that discussion.

This RFC is more at the 'feature request' stage rather than ready for an RFC PR. I would recommend the author close it for now and open either an issue on this repo or a thread on internals.rust-lang.org in order to gather the details required for an RFC.

**Comment by KalitaAlexey** [comment]
As @nrc suggested I closed this PR and opened a [thread on internals.rust-lang.org](https://internals.rust-lang.org/t/pre-rfc-allow-self-parameter-be-any-type/3739).

File: text/0000-any_type_self.md (hunk 3 above)
Comment (ticki) [inline text/0000-any_type_self.md]: But, how would you construct a value of `Vec<Self>` from a value of `Self`? Clearly, that doesn't make sense.

What you would need is a trait, like `From<Self>` which defines that `self` is constructed from the value in question, but at that point you introduce hidden conversions.

I believe that some form of generalized `AsRef` is what's needed.

Comment (KalitaAlexey) [inline text/0000-any_type_self.md]: @ticki I can't understand you.
*Type* depends on `Self` = *Type* has `Self` as generic parameter or associated type.

Comment (ticki) [inline text/0000-any_type_self.md]: Yes, but you need to define how to construct the value of that type when using the dot operator.

Comment (KalitaAlexey) [inline text/0000-any_type_self.md]: @ticki I don't quite understood you. What is *dot operator*?

**Pull request closed without merging**